comet/broker.py

```diff
@@ -482,33 +513,31 @@ async def gather_update(ts, roots):
     given roots that were registered after the given timestamp.
     """
     update = dict()
-    async with lock_datasets:
-        for root in roots:
+    for root in roots:
+        async with lock_datasets:
```

Comment:
the significance of these two being switched is that the lock is only held for the length of the asyncio.gather, but not for anything that follows? What is the lock required for?

Reply:
Correct. The lock protects access to the datasets, but also for the `roots_of_dataset*` I think.
Not locking for the whole time anymore means that there can be newer updates returned for some parts of the request. But it should be all good, because the update is tagged with a timestamp from before all this.